### PR TITLE
Popover: Avoid paint on popovers when scrolling

### LINKED ISSUE #45382: Components: Popover uses top and left positioning hurting scroll performance
Using top and left positioning triggers a repaint of the area which could be avoided using transform : translate(x, y), improving performances (currently the scroll performances of the editor are not really good, even more when using fixed backgrounds)

https://github.com/WordPress/gutenberg/blob/df98b4b46ab1951228b6021ce4e2b868ccd3848c/packages/components/src/popover/index.tsx#L466

https://user-images.githubusercontent.com/4660731/198627124-e3490bc7-91a0-467f-b6ab-4ba6d7a73f28.mov

### PR DESCRIPTION
**Pull request by corentin-gautier**
## What?

This PR changes the way popovers are positioned from top/left to translate(x, y) to avoid repaint on scroll

## Why?

Changing the top/left properties triggers browser repaint which will result in poor performances

## Testing Instructions

1. Open the dev console and enable the paint flashing (under rendering)
2. Select a block
3. Scroll so that the block toolbar needs to be re-positioned
4. There should not be paint on the toolbar

Fixes #45382

### Before
https://user-images.githubusercontent.com/4660731/198627124-e3490bc7-91a0-467f-b6ab-4ba6d7a73f28.mov

### After

https://user-images.githubusercontent.com/4660731/204805527-1a0b3cf3-4cd2-4a7f-8c1b-6464ea3a182e.mov

### CODE REVIEW

File: packages/components/src/popover/style.scss
Comment (tyxla) [inline packages/components/src/popover/style.scss:10]: See https://github.com/WordPress/gutenberg/pull/46197#discussion_r1036847433 for a suggestion regarding how we add / remove this CSS property.

File: packages/components/src/popover/index.tsx
Comment (tyxla) [inline packages/components/src/popover/index.tsx:499]: Previously, while animating we were accepting decimal values for `x` and `y`, and now we're rounding them. This is likely the reason for a bunch of e2e tests failing.

Comment (corentin-gautier) [reply packages/components/src/popover/index.tsx:499]: @tyxla Yes it was suggested by @ciampo here : https://github.com/WordPress/gutenberg/pull/45545#discussion_r1035127580

File: packages/components/src/popover/style.scss
Comment (corentin-gautier) [reply packages/components/src/popover/style.scss:10]: @tyxla since the number of popover is limited I don't think this is a big issue, BUT it might actually not be needed since popovers seem to get a translateZ for free 😄

File: packages/components/src/popover/index.tsx
Comment (corentin-gautier) [reply packages/components/src/popover/index.tsx:499]: @tyxla The tests indeed fail because of this but I have no idea how to change the expected values

File: packages/components/src/popover/style.scss
Comment (corentin-gautier) [reply packages/components/src/popover/style.scss:10]: @tyxla I have reverted the change since the translateZ does the job, however it seems to trigger repaint of some of the elements inside the block toolbar which is ... weird 😅

Comment (ciampo) [reply packages/components/src/popover/style.scss:10]: As @corentin-gautier mentioned, technically we shouldn't need it because of the `translateZ()` value that framer motion already adds — in that sense, I didn't expect the browser to trigger repaints as reported in the previous message 🤷 although browser heuristics about when to create a new composite layer can change and evolve in time, so we're never really guaranteed.

Regarding the when we should add/remove this prop, it's worth mentioning that [the `Popover` component should be rendered only when the popover is actually visible](https://github.com/WordPress/gutenberg/blob/f0fe8f64fe0afef0f383ffa8ea6320d1c774f600/packages/components/src/popover/README.md?plain=1#L11) — therefore, it seems OK to me that we keep that CSS rule as always ON: it's only rendered when the popover is open.

File: packages/components/src/popover/index.tsx
Comment (ciampo) [reply packages/components/src/popover/index.tsx:499]: `npm run test:unit:update` should update the snapshots — you should then commit those changes

**Comment by ellatrix** [comment]
Looks like this is making the toolbar a lot shakier when scrolling?

**Comment by corentin-gautier** [comment]
> Looks like this is making the toolbar a lot shakier when scrolling?

Having paint flashing enabled + the element selected in the dev tools makes it shaky, Without this it behaves normally 🙂

**Comment by jasmussen** [comment]
Feels pretty solid to me:

https://user-images.githubusercontent.com/1204802/205252348-c5b346b1-bf78-4d40-9f29-44baf05f0f65.mov

Thank you for all these PRs! 👏

File: packages/components/src/popover/style.scss
Comment (corentin-gautier) [reply packages/components/src/popover/style.scss:10]: @ciampo do you recommend i add it back then ?

File: packages/components/src/popover/index.tsx
Comment (corentin-gautier) [reply packages/components/src/popover/index.tsx:499]: @ciampo I ran it but there nothing to commit 🤔

File: packages/components/src/popover/style.scss
Comment (ciampo) [reply packages/components/src/popover/style.scss:10]: It would be good if you could investigate a bit more into why removing `will-change: transform` causes paint flashes in the toolbar, and see if we can achieve the desired result in any other way.

But if after the investigation the only guaranteed way to avoid paint flashes if by having `will-change: transform` always applied, then yes — I would recommend you add it back at that point

Comment (corentin-gautier) [reply packages/components/src/popover/style.scss:10]: @ciampo I looked into it yesterday but couldn't find the cause, it seems some of the svgs inside the toolbar trigger repaint when the toolbar position is updated ...

There are other tricks to avoid paint like using `backface-visibility: hidden;` so that's an option too :)

Comment (ciampo) [reply packages/components/src/popover/style.scss:10]: Have we tried setting `contain: paint` on the toolbar buttons, out of curiosity?

> There are other tricks to avoid paint like using backface-visibility: hidden; so that's an option too :)

Yup, although that doesn't really change the fact that we'd be "forcing" the popover to be on its own composite layer — so `will-change: transform` seems like a better (more semantic) choice

File: packages/components/src/popover/index.tsx
Comment (ciampo) [reply packages/components/src/popover/index.tsx:499]: Do you mind rebasing and fixing conflicts? I'm currently unable to see that tests are failing, which makes it harder to make an informed suggestion 😅

File: packages/components/src/popover/style.scss
Comment (corentin-gautier) [reply packages/components/src/popover/style.scss:10]: @ciampo using `contain: paint;` is .... worse 😂 It triggers paint of all the buttons (I know it's very weird ahah)

https://user-images.githubusercontent.com/4660731/205274637-33305c9f-8377-40ec-8ac2-4f898e058e51.mov

Comment (ciampo) [reply packages/components/src/popover/style.scss:10]: 🤷  Let's stick to `wil-change: transform` then! Thank you for the investigation :) 

If you ever feel like digging more into this and come up with more insights, I'm always eager to see how we can improve this component further!

**Review by tyxla — APPROVED** [review]
This looks and works well for me, thanks for the great work @corentin-gautier! 🚀

**Comment by ciampo** [comment]
Before merging, do you mind adding a CHANGELOG entry for the changes in the `@wordpress/components` package, as suggested by Aaron in [this other PR](https://github.com/WordPress/gutenberg/pull/46196#pullrequestreview-1200288177)?

We could actually add an entry also for #46197 which was merged without one

File: packages/components/src/popover/index.tsx
Comment (ciampo) [reply packages/components/src/popover/index.tsx:499]: @tyxla I can see unit tests erroring because of the `act` warning — do you mind helping out here? (I'm going to be AFK in the following weeks)

Comment (tyxla) [reply packages/components/src/popover/index.tsx:499]: Sure thing! Happy to aid with that next week.

**Comment by tyxla** [comment]
@corentin-gautier tests appear to be failing for a few reasons here:

* E2e tests - failing because the e2e tests expect `x` and `y` values to not be rounded, and we're now rounding them. It should be fine to manually adjust them to the new rounded values.
* Unit tests - failing because of the act warnings, seems to be because there's a promise running after the popovers are fully positioned, which makes the tests fragile. A straightforward way to fix them is to `unmount()` at the end of the failing tests, in order to cancel any pending effects that would cause `act()` warnings. 

Let me know if that helps!

**Comment by corentin-gautier** [comment]
> @corentin-gautier tests appear to be failing for a few reasons here:
> 
> * E2e tests - failing because the e2e tests expect `x` and `y` values to not be rounded, and we're now rounding them. It should be fine to manually adjust them to the new rounded values.
> * Unit tests - failing because of the act warnings, seems to be because there's a promise running after the popovers are fully positioned, which makes the tests fragile. A straightforward way to fix them is to `unmount()` at the end of the failing tests, in order to cancel any pending effects that would cause `act()` warnings.
> 
> Let me know if that helps!

I am not familiar at all with gutenberg's test system, I have tried searching for the values in the failing tests but didn't find any, where should I edit those values ?

**Comment by tyxla** [comment]
> I am not familiar at all with gutenberg's test system, I have tried searching for the values in the failing tests but didn't find any, where should I edit those values ?

Okay, let's take this step by step 😉 First, I see you rebased after merging the other PR, which I was going to ask you, but glad to see you took care of it 🙌 

On the unit tests failure, being honest, that's directly not related to your PR. It's related to the fact that we're not cleaning up the remaining running popover effects in time for the next test.  I've filed a fix for that in #46303, and after rebasing it, I believe that the unit tests should be more reliable and the changes here should no longer break them.

On the e2e failure, I just took a more thorough look, and it seems like those might actually indicate something unexpected going on, meaning that we might to further iron out things. 

The best way to understand what's going on is to run the tests locally. Here are the steps you need to take:

* Run the dev build (`npm run dev`)
* Install Docker if you haven't yet.
* Run Docker and setup wp-env locally (`npm run wp-env start`)
* Apply this diff to run only one of the failing tests:

```
diff --git a/test/e2e/specs/editor/blocks/paragraph.spec.js b/test/e2e/specs/editor/blocks/paragraph.spec.js
index c713379569..5f71fc299d 100644
--- a/test/e2e/specs/editor/blocks/paragraph.spec.js
+++ b/test/e2e/specs/editor/blocks/paragraph.spec.js
@@ -160,7 +160,7 @@ test.describe( 'Paragraph', () => {
                } );

                test.describe( 'Dragging positions', () => {
-                       test( 'Only the first block is an empty paragraph block', async ( {
+                       test.only( 'Only the first block is an empty paragraph block', async ( {
                                editor,
                                page,
                                draggingUtils,
```
* Run the e2e tests, particularly the suite with the failing test: `npm run test:e2e:playwright test/e2e/specs/editor/blocks/paragraph.spec.js`
* Witness the error and try to correct it. 

In this particular scenario, it seems like the popover changes introduced in this PR alter the dragging bounding box, but not the block bounding box, and these no longer match. We might need to make some additional changes and add the rounding to the actual block in the list of blocks as well. 

Hope that helps - let me know if you have any further questions or if you're blocked!

**Comment by corentin-gautier** [comment]
@tyxla Well ... that's awkward 😆 
Screenshot 2022-12-05 at 13 31 31

Screenshot 2022-12-05 at 13 46 53

**Comment by tyxla** [comment]
That is odd! Is there any chance you're running the e2e tests against the `trunk` and not against your branch? Noting that you actually have to build while on your branch before running the tests.

**Comment by corentin-gautier** [comment]
@tyxla Hum, yes there's a chance 😅 I now have the same result but not much more info ...

**Comment by corentin-gautier** [comment]
@tyxla I can fix the test by using a rounding function on `firstBlockBox`

Something like this :

```js
const roundBoundingBox = (obj) => {
	Object.keys(obj).forEach(k => {
		obj[k] = Math.round(obj[k]);
	});
	return obj;
}
```

But I don't know if this should be the way we fix this

File: test/e2e/specs/editor/blocks/paragraph.spec.js
Comment (tyxla) [inline test/e2e/specs/editor/blocks/paragraph.spec.js]: Sorry, but this just won't do the job. There's a reason why those two don't match - we're rounding the `popover` values and not the block values, so altering the test to "just work" because we're rounding the expected values won't do. We must seek the reason for this failure, e.g. why the block doesn't round the values, and solve the issue in its root instead of in the test.

**Review by tyxla — CHANGES_REQUESTED** [review]

**Comment by corentin-gautier** [comment]
@tyxla thing is : if we round the position of the popover, then it will never be strictly equal to the position of the block (which is calculated with a getBoundingClientRect and will return sub-pixel values, https://playwright.dev/docs/api/class-locator#locator-bounding-box)

From what I understand of the tests, their are testing that the dropzone (a BlockPopover) position is equal to the block (a paragraph) position.

**Comment by tyxla** [comment]
> @tyxla thing is : if we round the position of the popover, then it will never be strictly equal to the position of the block (which is calculated with a getBoundingClientRect and will return sub-pixel values, https://playwright.dev/docs/api/class-locator#locator-bounding-box)
> 
> From what I understand of the tests, their are testing that the dropzone (a BlockPopover) position is equal to the block (a paragraph) position.

Indeed, but then that points to a different problem - we either should change our expectations that the two positions will be equal (meaning, reconsidering the test completely instead of just rounding one of the values), or we should use the same heuristics - either both rounding or both non-rounding. Since rounding both is a no-go, should we reconsider the rounding for the popover position?

**Comment by corentin-gautier** [comment]
> Indeed, but then that points to a different problem - we either should change our expectations that the two positions will be equal (meaning, reconsidering the test completely instead of just rounding one of the values), or we should use the same heuristics - either both rounding or both non-rounding. Since rounding both is a no-go, should we reconsider the rounding for the popover position?

As I mentioned before it was suggested by @ciampo [here](https://github.com/WordPress/gutenberg/pull/45545#discussion_r1035127580)  in reference to [this](https://floating-ui.com/docs/misc#subpixel-and-accelerated-positioning)

> x and y can contain decimals, so unless the transform translation is placed evenly on the device's subpixel grid, then there will be blurring

I think it's reasonable to assume that the values won't be exactly the same (and by that I mean rounded, so not far from the same) because as you said we can't round the block's position but rounding the popover's position seems to be required in order to get a sharp UI

That being said it wasn't rounded before so maybe this could be the subject of another PR ?

**Comment by tyxla** [comment]
> That being said it wasn't rounded before so maybe this could be the subject of another PR ?

I'm inclined to agree with that. I'd love to get a second opinion from @mirka if possible.

**Comment by mirka** [comment]
> > That being said it wasn't rounded before so maybe this could be the subject of another PR ?
> 
> I'm inclined to agree with that. I'd love to get a second opinion from @mirka if possible.

You're asking whether we should leave the rounding part out of this PR? I wouldn't be opposed to that, but also it looks easy enough to alter the test assertions so it has tolerance for subpixel differences.

By the way, I think we could encapsulate the intent better if we explicitly tested for the tolerance, rather than apply rounding to the block values. So `expect(Math.abs(ax-bx)).toBeLessThan(1)` or something like that. Just a thought.

**Comment by tyxla** [comment]
> By the way, I think we could encapsulate the intent better if we explicitly tested for the tolerance, rather than apply rounding to the block values. So `expect(Math.abs(ax-bx)).toBeLessThan(1)` or something like that. Just a thought.

Loving that suggestion FWIW 👍 I'm happy to go with that - WDYT @corentin-gautier?

With regards to the failing unit tests - @jsnajdr [mentioned](https://github.com/WordPress/gutenberg/pull/46303#issuecomment-1340984200) he has a [branch](https://github.com/WordPress/gutenberg/tree/try/popover-positioning-classname) that could fix them - would you like to try cherry-picking them to your branch and seeing if it fixes the tests?

**Comment by jsnajdr** [comment]
After #46429 is merged, you can rebase this PR on top of it and it should solve the failing unit tests (that test popover positioning).

**Comment by corentin-gautier** [comment]
@tyxla @mirka what do you think of this : https://github.com/WordPress/gutenberg/pull/46187/commits/b50ca9fdc5db0a5b692a59278b3680a423a6407d ?

I've tried keeping the tests to be one liners instead of duplicating for x and x

**Review by tyxla — APPROVED** [review]
This looks great! 💯  Thanks @corentin-gautier 🙌 

A potential follow-up could be creating a custom inline matcher to make the assertion more eloquent, but that's not a big deal IMHO, since your comment is useful in making the intent clear.

I'll ship this for you 🚀 

Thanks again and keep up the great work 🥇

File: packages/components/src/popover/index.tsx
Comment (andrewserong) [reply packages/components/src/popover/index.tsx:499]: It looks like the change here results in `0` values not being applied to update popover positions for popovers that are intended to be flush with the edge of the screen. I've opened up a potential fix over in #51320.

**Pull request merged**
